### PR TITLE
[WIP] Global bs setup for lr scaling, partition global batch size across ranks

### PR DESCRIPTION
**Pull request by pzharrington**
Still in progress, but adjusting things so:
1. User sets the global batch size and a base batch size, and learning rate is scaled based on their ratio
2. Data loaders partition global batch across ranks so epoch time decreases if you add workers
    * Default data loader uses `DistributedSampler`
    * DALI data loader simply shortens pipe length

Still need to finish testing, some outlying issues:
1. Need to run more epochs when we scale up global batch size, so probably (?) want to hard-code that in the configs
2. The "epochs" are ill-defined because we are randomly sampling crops from the simulations. Thus the "length" of an epoch is just set by `Nsamples` and `Nsamples_val` -- these need to be at least as big as our biggest global batch size we plan on using. But adjusting them now will change the epoch times, e.g. Josh's sample output in the README profiling section. So probably will have some separate config with current `Nsamples` for the profiling runs, and then different ones for the DDP scaling runs.

Also added some stuff to README, but incomplete.

### CODE REVIEW

File: config/UNet.yaml [hunk 1]
```diff
@@ -4,13 +4,13 @@ base: &base
   weight_init: {conv_init: 'normal', conv_scale: 0.02, conv_bias: 0.}
   lambda_rho: 1E-2 # weight for additional rho loss term
   full_scale: True # whether or not to use all 6 of the scales in U-Net
-  batch_size: 1
+  batch_size: 64 # number of samples per training batch (global batch size for distributed training)
```
Comment (sparticlesteve) [inline config/UNet.yaml]: the parentheses statement makes it sound like this is global batch size. I'd just say that this is local batch size, and optionally add that global batch size is (local bs * num ranks).

Comment (pzharrington) [reply config/UNet.yaml]: That actually is global batch size. In `train.py` we set local batch size based on num ranks:
https://github.com/NERSC/sc21-dl-tutorial/blob/d2f4a8b4c551db915ca96a064e993217e9faea91/train.py#L224

Comment (sparticlesteve) [reply config/UNet.yaml]: Ah, ok, then this config can be used for strong scaling, but for weak-scaling we need a new config for each scale point, right?

Comment (pzharrington) [reply config/UNet.yaml]: Right, so sorta following last year's tutorial setup -- different config for each (global) batchsize

Comment (sparticlesteve) [reply config/UNet.yaml]: ok, thanks.

File: train_graph.py [hunk 2]
```diff
@@ -237,19 +237,25 @@ def train(params, args, local_rank, world_rank, world_size):
...
                       choices=['synthetic', 'inmem', 'lowmem', 'dali-lowmem'],
                       help="dataloader configuration. choices: 'synthetic', 'inmem', 'lowmem', 'dali-lowmem'")
+  parser.add_argument("--local_batch_size", default=None, type=int, help='local batchsize (manually override global_batch_size config setting)')
+  parser.add_argument("--num_epochs", default=None, type=int, help='number of epochs to run')
+  parser.add_argument("--num_data_workers", default=None, type=int, help='number of data workers for data loader')
   args = parser.parse_args()
-  
+
+  if (args.enable_benchy and args.enable_manual_profiling):
+      raise "Enable either benchy profiling or manual profiling, not both."
```
Comment (sparticlesteve) [inline train_graph.py:257]: You can't raise a string. Can you just wrap that string in a `RuntimeError(...)`?

**Comment by sparticlesteve** [comment]
Besides the exception thing, LGTM.

Let's at least try and get @romerojosh 's approval before merging.

File: train_graph.py (hunk 2 above)
Comment (romerojosh) [reply train_graph.py:257]: That is my code. I can fix it.

Comment (sparticlesteve) [reply train_graph.py:257]: we'll fix in a follow-up
